### PR TITLE
parser: Add 'SHOW BINARY LOG STATUS'

### LINKED ISSUE #48673: Support replacement for `SHOW MASTER STATUS`
## Enhancement

The `SHOW MASTER STATUS` syntax has been deprecated in [MySQL 8.2.0](https://dev.mysql.com/doc/relnotes/mysql/8.2/en/news-8-2-0.html#mysqld-8-2-0-sql-syntax). The new syntax is `SHOW BINARY LOG STATUS`.

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #48673 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
The `SHOW BINARY LOG STATUS` statement is now an alias for `SHOW MASTER STATUS`
```

### CODE REVIEW

**Comment by dveeden** [comment]
```
sql> SHOW MASTER STATUS;
+-------------+--------------------+--------------+------------------+-------------------+
| File        | Position           | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set |
+-------------+--------------------+--------------+------------------+-------------------+
| tidb-binlog | 445702800248668160 |              |                  |                   |
+-------------+--------------------+--------------+------------------+-------------------+
1 row in set (0.0011 sec)

sql> SHOW BINARY LOG STATUS;
+-------------+--------------------+--------------+------------------+-------------------+
| File        | Position           | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set |
+-------------+--------------------+--------------+------------------+-------------------+
| tidb-binlog | 445702801681022976 |              |                  |                   |
+-------------+--------------------+--------------+------------------+-------------------+
1 row in set (0.0012 sec)
```

**Comment by dveeden** [comment]
Looks like this change breaks the `LOG()` function

```
sql> SHOW MASTER STATUS;
+-------------+--------------------+--------------+------------------+-------------------+
| File        | Position           | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set |
+-------------+--------------------+--------------+------------------+-------------------+
| tidb-binlog | 445703060158414848 |              |                  |                   |
+-------------+--------------------+--------------+------------------+-------------------+
1 row in set (0.0010 sec)

sql> SHOW BINARY LOG STATUS;
+-------------+--------------------+--------------+------------------+-------------------+
| File        | Position           | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set |
+-------------+--------------------+--------------+------------------+-------------------+
| tidb-binlog | 445703060522795008 |              |                  |                   |
+-------------+--------------------+--------------+------------------+-------------------+
1 row in set (0.0012 sec)

sql> SELECT LOG(2);
ERROR: 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 11 near "(2)" 

```

**Comment by dveeden** [comment]
```
sql> SHOW BINARY LOG STATUS;
+-------------+--------------------+--------------+------------------+-------------------+
| File        | Position           | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set |
+-------------+--------------------+--------------+------------------+-------------------+
| tidb-binlog | 445704591635120128 |              |                  |                   |
+-------------+--------------------+--------------+------------------+-------------------+
1 row in set (0.0017 sec)

sql> SHOW MASTER STATUS;
+-------------+--------------------+--------------+------------------+-------------------+
| File        | Position           | Binlog_Do_DB | Binlog_Ignore_DB | Executed_Gtid_Set |
+-------------+--------------------+--------------+------------------+-------------------+
| tidb-binlog | 445704592180379648 |              |                  |                   |
+-------------+--------------------+--------------+------------------+-------------------+
1 row in set (0.0011 sec)

sql> SELECT LOG(2);
+--------------------+
| LOG(2)             |
+--------------------+
| 0.6931471805599453 |
+--------------------+
1 row in set (0.0015 sec)

sql> CREATE TABLE LOG(id SERIAL);
Query OK, 0 rows affected (0.0508 sec)
```

File: pkg/parser/parser_test.go [hunk 1]
```diff
@@ -1203,6 +1203,7 @@ func TestDBAStmt(t *testing.T) {
 		{`SHOW PROFILE CPU FOR QUERY 2 LIMIT 1,1`, true, "SHOW PROFILE CPU FOR QUERY 2 LIMIT 1,1"},
 		{`SHOW PROFILE CPU, MEMORY, BLOCK IO, CONTEXT SWITCHES, PAGE FAULTS, IPC, SWAPS, SOURCE FOR QUERY 1 limit 100`, true, "SHOW PROFILE CPU, MEMORY, BLOCK IO, CONTEXT SWITCHES, PAGE FAULTS, IPC, SWAPS, SOURCE FOR QUERY 1 LIMIT 100"},
 		{`SHOW MASTER STATUS`, true, "SHOW MASTER STATUS"},
+		{`SHOW BINARY LOG STATUS`, true, "SHOW MASTER STATUS"},
```
Comment (lance6716) [inline pkg/parser/parser_test.go:1206]: I slightly prefer we restore the two SQL to SHOW BINARY LOG STATUS, because SHOW MASTER STATUS is deprecated.

**Review by lance6716 — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]
It's better to confirm this change by the tidb tools maintainers, because they are the actual users of "SHOW MASTER STATUS"

**Comment by dveeden** [comment]
Holding to resolve the comments from @tiancaiamao and @lance6716

File: pkg/parser/parser_test.go (hunk 1 above)
Comment (dveeden) [reply pkg/parser/parser_test.go:1206]: This is possible, but needs more changes.

From `git diff -- . ':(exclude)pkg/parser/parser.go'`

```diff
diff --git a/pkg/executor/builder.go b/pkg/executor/builder.go
index d2a9e98e9a..a3761be1a3 100644
--- a/pkg/executor/builder.go
+++ b/pkg/executor/builder.go
@@ -875,7 +875,7 @@ func (b *executorBuilder) buildShow(v *plannercore.PhysicalShow) exec.Executor {
 		Extractor:             v.Extractor,
 		ImportJobID:           v.ImportJobID,
 	}
-	if e.Tp == ast.ShowMasterStatus {
+	if e.Tp == ast.ShowMasterStatus || e.Tp == ast.ShowBinlogStatus {
 		// show master status need start ts.
 		if _, err := e.Ctx().Txn(true); err != nil {
 			b.err = err
diff --git a/pkg/executor/show.go b/pkg/executor/show.go
index f4b9847a81..0ac276bbea 100644
--- a/pkg/executor/show.go
+++ b/pkg/executor/show.go
@@ -253,7 +253,7 @@ func (e *ShowExec) fetchAll(ctx context.Context) error {
 		return e.fetchShowPlugins()
 	case ast.ShowProfiles:
 		// empty result
-	case ast.ShowMasterStatus:
+	case ast.ShowMasterStatus, ast.ShowBinlogStatus:
 		return e.fetchShowMasterStatus()
 	case ast.ShowPrivileges:
 		return e.fetchShowPrivileges()
diff --git a/pkg/parser/ast/dml.go b/pkg/parser/ast/dml.go
index 7636ebf98d..69ea4efb3d 100644
--- a/pkg/parser/ast/dml.go
+++ b/pkg/parser/ast/dml.go
@@ -2978,6 +2978,7 @@ const (
 	ShowCreateResourceGroup
 	ShowImportJobs
 	ShowCreateProcedure
+	ShowBinlogStatus
 )
 
 const (
@@ -3067,6 +3068,8 @@ func (n *ShowStmt) Restore(ctx *format.RestoreCtx) error {
 
 	ctx.WriteKeyWord("SHOW ")
 	switch n.Tp {
+	case ShowBinlogStatus:
+		ctx.WriteKeyWord("BINARY LOG STATUS")
 	case ShowCreateTable:
 		ctx.WriteKeyWord("CREATE TABLE ")
 		if err := n.Table.Restore(ctx); err != nil {
diff --git a/pkg/parser/parser.y b/pkg/parser/parser.y
index e13d983883..1aac154278 100644
--- a/pkg/parser/parser.y
+++ b/pkg/parser/parser.y
@@ -11362,7 +11362,7 @@ ShowStmt:
 |	"SHOW" "BINARY" "LOG" "STATUS"
 	{
 		$$ = &ast.ShowStmt{
-			Tp: ast.ShowMasterStatus,
+			Tp: ast.ShowBinlogStatus,
 		}
 	}
 |	"SHOW" OptFull "PROCESSLIST"
diff --git a/pkg/parser/parser_test.go b/pkg/parser/parser_test.go
index 9eec7ee1df..17bfc31a78 100644
--- a/pkg/parser/parser_test.go
+++ b/pkg/parser/parser_test.go
@@ -1203,7 +1203,7 @@ func TestDBAStmt(t *testing.T) {
 		{`SHOW PROFILE CPU FOR QUERY 2 LIMIT 1,1`, true, "SHOW PROFILE CPU FOR QUERY 2 LIMIT 1,1"},
 		{`SHOW PROFILE CPU, MEMORY, BLOCK IO, CONTEXT SWITCHES, PAGE FAULTS, IPC, SWAPS, SOURCE FOR QUERY 1 limit 100`, true, "SHOW PROFILE CPU, MEMORY, BLOCK IO, CONTEXT SWITCHES, PAGE FAULTS, IPC, SWAPS, SOURCE FOR QUERY 1 LIMIT 100"},
 		{`SHOW MASTER STATUS`, true, "SHOW MASTER STATUS"},
-		{`SHOW BINARY LOG STATUS`, true, "SHOW MASTER STATUS"},
+		{`SHOW BINARY LOG STATUS`, true, "SHOW BINARY LOG STATUS"},
 		{`SHOW PRIVILEGES`, true, "SHOW PRIVILEGES"},
 		// for show character set
 		{"show character set;", true, "SHOW CHARSET"},
diff --git a/pkg/planner/core/planbuilder.go b/pkg/planner/core/planbuilder.go
index 9f5adbca51..8643d8649e 100644
--- a/pkg/planner/core/planbuilder.go
+++ b/pkg/planner/core/planbuilder.go
@@ -5363,6 +5363,9 @@ func buildShowSchema(s *ast.ShowStmt, isView bool, isSequence bool) (schema *exp
 	var names []string
 	var ftypes []byte
 	switch s.Tp {
+	case ast.ShowBinlogStatus:
+		names = []string{"File", "Position", "Binlog_Do_DB", "Binlog_Ignore_DB", "Executed_Gtid_Set"}
+		ftypes = []byte{mysql.TypeVarchar, mysql.TypeLonglong, mysql.TypeVarchar, mysql.TypeVarchar, mysql.TypeVarchar}
 	case ast.ShowProcedureStatus, ast.ShowFunctionStatus:
 		return buildShowProcedureSchema()
 	case ast.ShowTriggers:
diff --git a/pkg/planner/core/planbuilder_test.go b/pkg/planner/core/planbuilder_test.go
index 81ee1e7bac..365fcccf5d 100644
--- a/pkg/planner/core/planbuilder_test.go
+++ b/pkg/planner/core/planbuilder_test.go
@@ -49,6 +49,7 @@ type visit struct {
 func TestShow(t *testing.T) {
 	node := &ast.ShowStmt{}
 	tps := []ast.ShowStmtType{
+		ast.ShowBinlogStatus,
 		ast.ShowEngines,
 		ast.ShowDatabases,
 		ast.ShowTables,
```

@lance6716  Should I push this?

Comment (dveeden) [reply pkg/parser/parser_test.go:1206]: Another option would be to restore both `SHOW MASTER STATUS` and `SHOW BINARY LOG STATUS` as the new `SHOW BINARY LOG STATUS`, but that might lead to compatibility issues.

Comment (lance6716) [reply pkg/parser/parser_test.go:1206]: Yes, I prefer you push it.

**Comment by lance6716** [comment]
And still need to request an approve from pkg/parser/OWNERS

File: pkg/parser/parser_test.go (hunk 1 above)
Comment (dveeden) [reply pkg/parser/parser_test.go]: Pushed as 01ff455beb256f7f5d9f6965bf9bc2d6b1f2a127

**Review by 3pointer — APPROVED** [review]

**Comment by 3pointer** [comment]
> It's better to confirm this change by the tidb tools maintainers, because they are the actual users of "SHOW MASTER STATUS"

looks like the compatibility of `SHOW MASTER STATUS` doesn't change. so It's okay for tidb-tools.

**Pull request merged**